lib/collection/src/operations/verification/query.rs

```diff
@@ -35,6 +35,52 @@ impl Query {
         }
         Ok(())
     }
+
+    /// Check that the query does not perform a fullscan based on the collection configuration.
+    async fn check_fullscan(
+        &self,
+        using: &str,
+        collection: &Collection,
+        strict_mode_config: &StrictModeConfig,
+    ) -> CollectionResult<()> {
+        // Check only applies on `unindexed_filtering_retrieve`
+        if strict_mode_config.unindexed_filtering_retrieve == Some(false) {
+            if let Query::Vector(_) = &self {
+                let config = collection.collection_config.read().await;
+
+                // ignore sparse vectors
+                let query_targets_sparse = config
+                    .params
+                    .sparse_vectors
+                    .as_ref()
+                    .is_some_and(|sparse| sparse.contains_key(using));
+                if query_targets_sparse {
+                    // sparse vectors are always indexed
+                    return Ok(());
+                }
+
+                // check HNSW configuration for vector
+                let vector_hnsw_config = &config
+                    .params
+                    .vectors
+                    .get_params(using)
+                    .and_then(|param| param.hnsw_config.as_ref());
+
+                let vector_hnsw_m = vector_hnsw_config.and_then(|hnsw| hnsw.m);
+                let vector_hnsw_payload_m = vector_hnsw_config.and_then(|hnsw| hnsw.payload_m);
```

Comment:
Payload_m is only an excuse if there is a filter by tenant/principal

Reply:
Got it, removing it from the check and will tackle multitenancy in a different PR.